log/loki.go

```diff
@@ -0,0 +1,419 @@
+// TODO clean old entries after push?
+func (h *lokiHook) loop() {
+	var (
+		msgs       = make([]tmpMsg, h.limit)
+		msgsToPush = make([]tmpMsg, h.limit)
+		dropped    int
+		count      int
+		ticker     = time.NewTicker(h.pushPeriod)
+		pushCh     = make(chan chan struct{})
+	)
+
+	defer close(pushCh)
+
+	go func() {
+		oldLogs := make([]tmpMsg, 0, h.limit*2)
+		for ch := range pushCh {
+			msgsToPush, msgs = msgs, msgsToPush
+			oldCount, oldDropped := count, dropped
+			count, dropped = 0, 0
+			close(ch) // signal that more buffering can continue
+
+			copy(oldLogs[len(oldLogs):len(oldLogs)+oldCount], msgsToPush[:oldCount])
+			oldLogs = oldLogs[:len(oldLogs)+oldCount]
+			n, err := h.push(oldLogs, oldDropped)
+			_ = err // TODO print it on terminal ?!?
+			if n > len(oldLogs) {
+				oldLogs = oldLogs[:0]
+				continue
+			}
+			oldLogs = oldLogs[:copy(oldLogs, oldLogs[n:])]
+		}
+	}()
+
+	for {
+		select {
+		case entry, ok := <-h.ch:
+			if !ok {
+				// TODO push one last time ?
+				return
+			}
+
+			if count == h.limit {
+				dropped++
+				continue
+			}
+
+			// Arguably we can directly generate the final marshalled version of the labels right here
+			// through sorting the entry.Data, removing additionalparams from it and then dumping it
+			// as the final marshal and appending level and h.labels after it.
+			// If we reuse some kind of big enough `[]byte` buffer we can also possibly skip on some
+			// of allocation. Combined with the cutoff part and directly pushing in the final data
+			// type this can be really a lot faster and to use a lot less memory
+			labels := make(map[string]string, len(entry.Data)+1)
+			for k, v := range entry.Data {
+				// TODO filter the additional ones
+				labels[k] = fmt.Sprint(v) // TODO optimize ?
+			}
+			for _, params := range h.labels {
+				labels[params[0]] = params[1]
+			}
+			labels["level"] = entry.Level.String()
+			// TODO have the cutoff here ?
+			// if we cutoff here we can cut somewhat on the backbuffers and optimize the inserting
+			// in/creating of the final Streams that we push
+			msgs[count] = tmpMsg{
+				labels: labels,
+				msg:    entry.Message,
+				t:      entry.Time.UnixNano(),
+			}
+			count++
+		case <-ticker.C:
+			ch := make(chan struct{})
+			pushCh <- ch
+			<-ch
+		}
+	}
+}
+
+func (h *lokiHook) push(msgs []tmpMsg, dropped int) (int, error) {
+	if len(msgs) == 0 {
+		return 0, nil
+	}
+
+	// TODO using time.Before was giving a lot of out of order, but even now, there are some, if the
+	// limit is big enough ...
+	sort.Slice(msgs, func(i, j int) bool {
+		return msgs[i].t < msgs[j].t
+	})
+
+	// We can technically cutoff during the addMsg phase, which will be even better if we sort after
+	// it as well ... maybe
+	cutOff := time.Now().Add(-time.Millisecond * 500).UnixNano() // probably better to be configurable
+
+	cutOffIndex := sort.Search(len(msgs), func(i int) bool {
+		return !(msgs[i].t < cutOff)
+	})
+
+	if cutOffIndex > len(msgs) {
```

Comment:
This can't happen, right? According to the [docs](https://golang.org/pkg/sort/#Search):

> Search returns the first true index. If there is no such index, Search returns n

So I think you can safely use `cutOffIndex`as returned.

Reply:
I guess so ...